### PR TITLE
Ouroboros C++ Core System

### PR DESCRIPTION
**Pull request by cjchanx**
Will be adding comments myself for errors I notice when reading code on phone.

This will be 
**Andromeda v4.0.0 (C++)**

**Commit Message**
Andromeda v4.0.0 (C++ Core System) [#47]
< Autogenerated things >

TODO List:

- [x] ~Remember to make a tag for the current repo as "Andromeda V3.3 C" or something like that before merge.~ 《Done》
- [x] ~ioc file shouldn't be changed much from Andromeda 3.3 (might be from Andromeda V3.2). Consider just grabbing the one from master.~ 《Done》
- [x] ~Increase the FreeRTOS heap4 size~ 《Done》

Comments:

### CODE REVIEW

File: Components/Core/Mutex.cpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+/**
+ * @brief This function is used to lock the Mutex.
+ * @param timeout_ms The time to wait for the Mutex before it fails. If timeout_ms is not provided, the function will wait indefinitely.
+ * @return True on success, false on failure.
+*/
+bool Mutex::Lock(uint32_t timeout_ms)
+{
+	return xSemaphoreTake(rtSemaphoreHandle, timeout_ms) == pdTRUE;
```
Comment (cjchanx) [inline Components/Core/Mutex.cpp]: Make sure this is in milliseconds (need to div by portTICK possibly)

File: Components/FlightControl/Inc/FlightTask.hpp [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+ * Description        : Primary flight task, default task for the system.
+ ******************************************************************************
+*/
+#ifndef SOAR_FLIGHTTASK_HPP_
+#define SOAR_FLIGHTTASK_HPP_
+#include "Task.hpp"
+#include "SystemDefines.hpp"
+
+class FlightTask : Task
+{
```
Comment (cjchanx) [inline Components/FlightControl/Inc/FlightTask.hpp:13]: Needs to be public inheritance to allow access to Task:: functions

File: Components/Communication/UARTTask.cpp [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+			(UBaseType_t)UART_TASK_PRIORITY,
+			(TaskHandle_t*)&rtTaskHandle);
+
+	SOAR_ASSERT(rtValue == pdPASS, "UARTTask::InitTask() - xTaskCreate() failed");
+
+	// Configure UART
+
+}
+
+void UARTTask::Run(void * pvParams)
```
Comment (AaronL11) [inline Components/Communication/UARTTask.cpp:48]: Definitely need some documentation here.

Comment (cjchanx) [reply Components/Communication/UARTTask.cpp:48]: Added: c56ea4d7748a2b77203be40b80fdfb6fe2847c53

File: Components/Core/Command.cpp [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+#include <cstring> 	// Support for memcpy
+/**
+ ******************************************************************************
+ * File Name          : Task.cpp
+ * Description        : Task contains the core component base class for all tasks.
+ ******************************************************************************
+*/
+
+/* Static Variable Init ------------------------------------------------------------------*/
+uint16_t Command::statAllocationCounter = 0; // Static variable init
```
Comment (cjchanx) [inline Components/Core/Command.cpp]: Make this explicitly atomic if possible

File: Components/Utils.cpp [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+uint32_t Utils::getCRC32(uint8_t* data, uint32_t size)
+{
+	// Generate a buffer padded to uint32_t
+    uint8_t pad = size % 4;
+    uint32_t buffer[(size + pad) / 4];
+    uint8_t* tempPtr = (uint8_t*)(&buffer[0]);
+
+    // Bytewise copy and pad
+	memcpy(tempPtr, data, size);
+	memset(tempPtr + size, 0, pad);
```
Comment (AgamAulakh) [inline Components/Utils.cpp:86]: math? size to memset = 4 - pad

**Review by AgamAulakh — CHANGES_REQUESTED** [review]

File: Components/Utils.cpp (hunk 5 above)
Comment (cjchanx) [reply Components/Utils.cpp:86]: Fixed: 569d772d9a80de0514082e2eb6241680be56b672

**Review by AgamAulakh — APPROVED** [review]

**Pull request merged**